staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/values.go

```diff
@@ -0,0 +1,571 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/common/types/traits"
+
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	"k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model"
+	"k8s.io/kube-openapi/pkg/validation/strfmt"
+)
+
+// UnstructuredToVal converts a Kubernetes unstructured data element to a CEL Val.
+func UnstructuredToVal(unstructured interface{}, schema *structuralschema.Structural) ref.Val {
+	if schema.XIntOrString {
+		switch unstructured.(type) {
+		case string, int64:
+			return types.DefaultTypeAdapter.NativeToValue(unstructured)
+		}
+		return types.NewErr("invalid data, expected XIntOrString value to be either a string or integer")
+	}
+	if schema.Type == "object" {
+		m, ok := unstructured.(map[string]interface{})
+		if !ok {
+			types.NewErr("invalid data, expected map[string]interface{} to match the provided schema with type=object")
+		}
+		if schema.XEmbeddedResource {
+			// only allow access to name and generateNames of metadata by pruning all other metadata fields.
+			return &unstructuredMap{
+				value:      m,
+				schema:     model.WithTypeAndObjectMeta(schema),
+				propSchema: func(key string) *structuralschema.Structural { schema := schema.Properties[key]; return &schema },
```

Comment:
we're setting `unstructuredMap.schema` to the augmented `model.WithTypeAndObjectMeta(schema)`, but the `unstructuredMap.propSchema` accessor is still only looking things up in the original unaugmented `schema.Properties` map... will this correctly provide access to type/metadata properties if they weren't explicitly in `schema`?

Reply:
Turns out this is a redundant call to WithTypeAndObjectMeta because we already call WithTypeAndObjectMeta in cel/validator.go before calling UnstructuredToVal. I've simplified the code and added a comment.